code/go/0chain.net/smartcontract/storagesc/allocation.go

```diff
@@ -180,93 +213,58 @@ func (sc *StorageSmartContract) newAllocationRequestInternal(
 	balances chainstate.StateContextI,
 	timings map[string]time.Duration,
 ) (resp string, err error) {
-	m := Timings{timings: timings, start: time.Now()}
-
+	m := Timings{timings: timings, start: common.ToTime(txn.CreationDate)}
 	var request newAllocationRequest
-	logging.Logger.Debug("new_allocation_request", zap.String("request", string(input)))
 	if err = request.decode(input); err != nil {
 		logging.Logger.Error("new_allocation_request_failed: error decoding input",
 			zap.String("txn", txn.Hash),
 			zap.Error(err))
 		return "", common.NewErrorf("allocation_creation_failed",
 			"malformed request: %v", err)
 	}
-
-	m.tick("decode")
-	if len(request.Blobbers) < (request.DataShards + request.ParityShards) {
-		logging.Logger.Error("new_allocation_request_failed: input blobbers less than requirement",
-			zap.String("txn", txn.Hash),
-			zap.Int("request blobbers", len(request.Blobbers)),
-			zap.Int("data shards", request.DataShards),
-			zap.Int("parity_shards", request.ParityShards))
-		return "", common.NewErrorf("allocation_creation_failed",
-			"Blobbers provided are not enough to honour the allocation")
+	if err := request.validate(common.ToTime(txn.CreationDate), conf); err != nil {
+		return "", common.NewErrorf("allocation_creation_failed", err.Error())
 	}
 
-	//if more than limit blobbers sent, just cut them
-	if len(request.Blobbers) > conf.MaxBlobbersPerAllocation {
-		logging.Logger.Error("new_allocation_request_failed: request blobbers more than max_blobbers_per_allocation",
-			zap.String("txn", txn.Hash),
-			zap.Int("requested blobbers", len(request.Blobbers)),
-			zap.Int("max blobbers per allocation", conf.MaxBlobbersPerAllocation))
-		logging.Logger.Info("Too many blobbers selected, max available", zap.Int("max_blobber_size", conf.MaxBlobbersPerAllocation))
-		request.Blobbers = request.Blobbers[:conf.MaxBlobbersPerAllocation]
+	if request.Owner == "" {
+		request.Owner = txn.ClientID
+		request.OwnerPublicKey = txn.PublicKey
 	}
 
-	inputBlobbers := sc.getBlobbers(request.Blobbers, balances)
-	if len(inputBlobbers.Nodes) < (request.DataShards + request.ParityShards) {
+	inputBlobbers := getBlobbersByIDs(request.Blobbers, balances)
+	if len(inputBlobbers) < (request.DataShards + request.ParityShards) {
 		logging.Logger.Error("new_allocation_request_failed: blobbers fetched are less than requested blobbers",
 			zap.String("txn", txn.Hash),
-			zap.Int("fetched blobbers", len(inputBlobbers.Nodes)),
+			zap.Int("fetched blobbers", len(inputBlobbers)),
 			zap.Int("data shards", request.DataShards),
 			zap.Int("parity_shards", request.ParityShards))
 		return "", common.NewErrorf("allocation_creation_failed",
-			"Not enough provided blobbers found in mpt")
+			"Blobbers provided are not enough to honour the allocation")
 	}
-
-	if request.Owner == "" {
-		request.Owner = txn.ClientID
-		request.OwnerPublicKey = txn.PublicKey
+	spMap := getStakePoolsByIDs(request.Blobbers, balances)
```

Comment:
getStakePoolsByIDs need to check if err is 'util.NodeNotFound' and return if any to avoid state hash mismatch error.

Reply:
@peterlimg not sure I understand, do you mean that inside `getStakePoolsByIDs` in the goroutine it is not checked? if getting from the pool there return NodeNotFound error, then it will be only logged and swallowed, is it what you meant?

Reply:
There was a problem with getBlobbersByIDs that errors were potentially getting lost. This was fixed in d2b3472aad5f6be5589a999f01fdf0b2d77c17d6. I assume that was @peterlimg's point.

Reply:
@Sriep yes, that's what I mean, but the fix in d2d3472 could only fix part of it, we should not push the `util.ValueNotPreent` to the errChan, as it would shadow the `util.NodeNotFound` error if any. See https://github.com/0chain/0chain/blob/69036192db4bc0a6dcf3b47aea8cf9990606f60d/code/go/0chain.net/smartcontract/storagesc/allocation.go#L613-L620 for how it was done in PR 1594.